test/integration/consul-container/test/gateways/gateway_endpoint_test.go

```diff
@@ -0,0 +1,250 @@
+	libcluster "github.com/hashicorp/consul/test/integration/consul-container/libs/cluster"
+	libservice "github.com/hashicorp/consul/test/integration/consul-container/libs/service"
+	"github.com/hashicorp/consul/test/integration/consul-container/libs/utils"
+	"github.com/hashicorp/go-cleanhttp"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"io"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+)
+
+var (
+	checkTimeout  = 1 * time.Minute
+	checkInterval = 1 * time.Second
+)
+
+// Creates a gateway service and tests to see if it is routable
+func TestAPIGatewayCreate(t *testing.T) {
+	t.Skip()
+	if testing.Short() {
+		t.Skip("too slow for testing.Short")
+	}
+
+	t.Parallel()
+
+	listenerPortOne := 6000
+
+	cluster := createCluster(t, listenerPortOne)
+
+	client := cluster.APIClient(0)
+
+	//setup
+	apiGateway := &api.APIGatewayConfigEntry{
+		Kind: "api-gateway",
+		Name: "api-gateway",
+		Listeners: []api.APIGatewayListener{
+			{
+				Port:     listenerPortOne,
+				Protocol: "tcp",
+			},
+		},
+	}
+	_, _, err := client.ConfigEntries().Set(apiGateway, nil)
+	assert.NoError(t, err)
+
+	tcpRoute := &api.TCPRouteConfigEntry{
+		Kind: "tcp-route",
+		Name: "api-gateway-route",
+		Parents: []api.ResourceReference{
+			{
+				Kind: "api-gateway",
+				Name: "api-gateway",
+			},
+		},
+		Services: []api.TCPService{
+			{
+				Name: libservice.StaticServerServiceName,
+			},
+		},
+	}
+
+	_, _, err = client.ConfigEntries().Set(tcpRoute, nil)
+	assert.NoError(t, err)
+
+	// Create a client proxy instance with the server as an upstream
+	_, gatewayService := createServices(t, cluster, listenerPortOne)
+
+	//check statuses
+	gatewayReady := false
+	routeReady := false
+
+	//make sure the gateway/route come online
+	require.Eventually(t, func() bool {
+		entry, _, err := client.ConfigEntries().Get("api-gateway", "api-gateway", nil)
+		assert.NoError(t, err)
+		apiEntry := entry.(*api.APIGatewayConfigEntry)
+		gatewayReady = isAccepted(apiEntry.Status.Conditions)
+
+		e, _, err := client.ConfigEntries().Get("tcp-route", "api-gateway-route", nil)
+		assert.NoError(t, err)
+		routeEntry := e.(*api.TCPRouteConfigEntry)
+		routeReady = isBound(routeEntry.Status.Conditions)
+
+		return gatewayReady && routeReady
+	}, time.Second*10, time.Second*1)
+
+	port, err := gatewayService.GetPort(listenerPortOne)
+	assert.NoError(t, err)
+	libassert.HTTPServiceEchoes(t, "localhost", port, "")
+}
+
+func isAccepted(conditions []api.Condition) bool {
+	return conditionStatusIsValue("Accepted", "True", conditions)
+}
+
+func isBound(conditions []api.Condition) bool {
+	return conditionStatusIsValue("Bound", "True", conditions)
+}
+
+func conditionStatusIsValue(typeName string, statusValue string, conditions []api.Condition) bool {
+	for _, c := range conditions {
+		if c.Type == typeName && c.Status == statusValue {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
We'll probably want to allow this to eventually target a specific `ResourceRef` instead of the first condition of `typeName`, but fine for now.